ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,851 @@
+    public boolean accept(File dir, String name) {
+      return name.endsWith(SUFFIX);
+    }
+  };
+
+  private static final long WAIT_TIME_FOR_CLEANUP_MS = 5 * Time.MsPerSec;
+  private static final String TEMP_LOG_SEGMENT_NAME_SUFFIX = BlobStore.SEPARATOR + "temp";
+
+  private final File dataDir;
+  private final String storeId;
+  private final StoreKeyFactory storeKeyFactory;
+  private final StoreConfig config;
+  private final StoreMetrics metrics;
+  private final Log srcLog;
+  private final DiskIOScheduler diskIOScheduler;
+  private final ScheduledExecutorService scheduler;
+  private final MessageStoreRecovery recovery;
+  private final Time time;
+  private final UUID sessionId;
+  private final UUID incarnationId;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private volatile boolean isActive = false;
+  private PersistentIndex srcIndex;
+
+  private Log tgtLog;
+  private PersistentIndex tgtIndex;
+  private long numSwapsUsed;
+  private StoreFindToken recoveryStartToken = null;
+  private CompactionLog compactionLog;
+  private volatile CountDownLatch runningLatch = new CountDownLatch(0);
+
+  /**
+   * Constructs the compactor component.
+   * @param dataDir the directory where all compactions need to be run.
+   * @param storeId the unique ID of the store.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to generate {@link StoreKey} instances.
+   * @param config the {@link StoreConfig} that defines configuration parameters.
+   * @param metrics the {@link StoreMetrics} to use to record metrics.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to schedule I/O.
+   * @param srcLog the {@link Log} to copy data from.
+   * @param scheduler the {@link ScheduledExecutorService} to use.
+   * @param recovery the {@link MessageStoreRecovery} to use to recover the index and log.
+   * @param time the {@link Time} instance to use.
+   * @param sessionId the sessionID of the store.
+   * @param incarnationId the incarnation ID of the store.
+   */
+  BlobStoreCompactor(String dataDir, String storeId, StoreKeyFactory storeKeyFactory, StoreConfig config,
+      StoreMetrics metrics, DiskIOScheduler diskIOScheduler, Log srcLog, ScheduledExecutorService scheduler,
+      MessageStoreRecovery recovery, Time time, UUID sessionId, UUID incarnationId) {
+    this.dataDir = new File(dataDir);
+    this.storeId = storeId;
+    this.storeKeyFactory = storeKeyFactory;
+    this.config = config;
+    this.metrics = metrics;
+    this.srcLog = srcLog;
+    this.diskIOScheduler = diskIOScheduler;
+    this.scheduler = scheduler;
+    this.recovery = recovery;
+    this.time = time;
+    this.sessionId = sessionId;
+    this.incarnationId = incarnationId;
+  }
+
+  /**
+   * If a compaction was in progress during a crash/shutdown, fixes the state so that the store is loaded correctly
+   * and compaction can resume smoothly on a call to {@link #resumeCompaction()}. Expected to be called before the
```

Comment:
This will be enforced in the `BlobStore` - Since both you and @nsivabalan have questions around the startup flow and the usage of the APIs, I will add startup code in a different commit.